typegen/evaluation/file_type_hints_collector.py

```diff
@@ -0,0 +1,159 @@
+import os.path
+import pathlib
+from typing import Dict
+import pandas as pd
+import libcst as cst
+from libcst.metadata import PositionProvider
+import constants
+from tracing import TraceDataCategory
+
+
+class FileTypeHintsCollector:
+    """Collects the type hints of multiple .py files."""
+    def __init__(self, project_dir: pathlib.Path):
+        self.project_dir = project_dir
+        self.typehint_data = pd.DataFrame()
+        self.file_path = ""
+
+    def collect_data(self, file_paths: list[pathlib.Path]) -> None:
+        """Collects the type hints of the provided file paths."""
+        for file_path in file_paths:
+            with file_path.open() as file:
+                file_content = file.read()
+
+            module = cst.parse_module(source=file_content)
+            module_and_meta = cst.MetadataWrapper(module)
+            relative_path = str(pathlib.Path(os.path.relpath(file_path.resolve(), self.project_dir.resolve())))
+            visitor = _TypeHintVisitor(relative_path)
+            module_and_meta.visit(visitor)
+            typehint_data = visitor.typehint_data.astype(constants.TraceData.TYPE_HINT_SCHEMA)
+
+            self.typehint_data = pd.concat(
+                [self.typehint_data, typehint_data], ignore_index=True
+            ).astype(constants.TraceData.TYPE_HINT_SCHEMA)
+
+
+class _TypeHintVisitor(cst.CSTVisitor):
+    METADATA_DEPENDENCIES = (PositionProvider,)
+
+    def __init__(self, file_path: str) -> None:
+        super().__init__()
+        self.file_path = file_path
+        self.typehint_data = pd.DataFrame(columns=constants.TraceData.TYPE_HINT_SCHEMA.keys())
+        self._scope_stack: list[cst.FunctionDef | cst.ClassDef] = []
+        self.imports: Dict[str, str] = {}
+
+    def _innermost_class(self) -> cst.ClassDef | None:
+        fromtop = reversed(self._scope_stack)
+        classes = filter(lambda p: isinstance(p, cst.ClassDef), fromtop)
+
+        first: cst.ClassDef | None = next(classes, None)  # type: ignore
+        return first
+
+    def _innermost_function(self) -> cst.FunctionDef | None:
+        fromtop = reversed(self._scope_stack)
+        fdefs = filter(lambda p: isinstance(p, cst.FunctionDef), fromtop)
+
+        first: cst.FunctionDef | None = next(fdefs, None)  # type: ignore
+        return first
+
+    def visit_ImportFrom(self, node: cst.ImportFrom) -> bool | None:
+        module_name: str = str(node.module.value)  # type: ignore
+        for name in node.names:  # type: ignore
+            class_name: str = str(name.name.value)
```

Comment:
No conversion should be necessary. value is a `str`

Reply:
Fixed

Reply:
Then add an `assert isinstance` statement, so that we get a clear error traceback if something did go wrong

Reply:
Done

Reply:
Ah, I have found the reason.
`value` is always a `str`, but `module` can be either an `cst.Attribute`, or a `cst.Name`.
`cst.Attribute` is likely when the case `from x.y import z` occurs; `x.y` being the attribute. This must be tested and fixed prior to merging

Reply:
Ah, this was fixed outside of the displayed diff. closing